File: LeetCode/Problems/239. Sliding Window Maximum/Solution2.js
// mono queue
var maxSlidingWindow = function(nums, k) {
    const n = nums.length;
    const q = [], res = [];
    for (let i = 0; i < k; ++i) {
        while (q.length > 0 && nums[q.at(-1)] <= nums[i]) q.pop();
        q.push(i);
    }
    res.push(nums[q[0]]);
    for (let i = k; i < n; ++i) {
        while (q.length > 0 && nums[q.at(-1)] <= nums[i]) q.pop();
        q.push(i);
        while (q[0] <= i - k) q.shift();
        res.push(nums[q[0]]);
    }
    return res;
};

File: LeetCode/Problems/239. Sliding Window Maximum/Solution2.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        const int n = nums.size();
        deque<int> q;
        vector<int> res;
        for (int i = 0; i < k; ++i) {
            while (q.size() && nums[q.back()] <= nums[i]) q.pop_back();
            q.push_back(i);
        }
        res.push_back(nums[q.front()]);
        for (int i = k; i < n; ++i) {
            while (q.size() && q.front() <= i - k) q.pop_front();
            while (q.size() && nums[q.back()] <= nums[i]) q.pop_back();
            q.push_back(i);
            res.push_back(nums[q.front()]);
        }
        return res;
    }
};
